contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,187 @@
+of the valid priority names are defined by a system wide mapping that maps each
+string to an integer. The PriorityName specified in a pod spec must be found in
+this map or the pod creation request will be rejected. If PriorityName is
+empty, "tier3" (see below) is assumed. I am open to other suggestions about
+what the default priority should be. Once the PriorityName is resolved to an
+integer, it is placed in "Priority" field of PodStatus.
+
+
+```
+type PodSpec struct {
+  ...
+  PriorityName string
+}
+
+type PodStatus struct {
+  ...
+  Priority int  // Populated by Admission Controller.
+}
+```
+
+### Priority names 
+
+A Kubernetes cluster has four predefined (built-in) priority names. The top
+priority is reserved for the system pods and its value is the largest 32-bit
+integer. The following list is an example of how the priorities and their
+values may look like. I am open to suggestions for the names and values of the
+priorities.
+
+
+```
+system  2147483647 (int_max)
+tier1   4000
+tier2   2000
+tier3   1000
+```
+
+
+### Resolving priority names 
+
+User requests sent to Kubernetes may have "PriorityName"  in their PodSpec.
+Admission controller resolves a PriorityName to its corresponding number and
+populates the "Priority" field of the pod status. The rest of Kubernetes
+components look at the "Priority" field of pod status and work with the integer
+value. In other words, "PriorityName" will be ignored by the rest of the
+system.
+
+### Modifying priority names 
+
+Admission controller can optionally receive a ConfigMap from the API server
+that defines a list of new priority names and their numbers or changes the
+integral value of existing ones. New configurations cannot introduce priority
+names mapped to a value higher than "system" and cannot change the value of
+"system" priority. The reason for this restriction is that kubernetes critical
+system pods will have "system" priority and no pod should be able to preempt
+them.
+
+Alternatively, we may ship Kubernetes with only the highest priority level
+hard-coded. The rest of the priority names are defined in a ConfigMap which
+could be shipped with Kubernetes.
+
+### Ordering of priorities 
+
+As mentioned earlier, a PriorityName is resolved by the admission controller to
+its integral value and Kubernetes components use the integral value. The higher
+the value, the higher the priority.
+
+### Dynamic change of priority names 
+
+Kubernetes supports change of priority names at runtime, however the changes
+won't change the priority of existing pods. As mentioned earlier, admission
+controller resolves priority names and populates the "Priority" field of pod
+specs. A new ConfigMap may be fed to the admission controller to define new
+priority names or change the integer value of existing names (other than
+"system"). These new changes affects only the pods created after the changes
+and existing pods will keep their existing Priority.
```

Comment:
@wojtek-t We can add a restriction on allowing deletion only when there is no pod with that priority class in the system, but completely disallowing deletion of priority classes wouldn't be practical IMO. Over time these classes might get accumulated and there would be no way to clean them up.
Race condition will exist when performing the check till the class is actually deleted, but I guess it will be OK given that deletion is probably infrequent and the fact that if a pod manages to sneak in due a race, it won't break the cluster.